### PR TITLE
erts: Fix fragmented send to finish before exiting

### LINKED ISSUE #5876: Binary memory leak in drv_binary associated with the dist port
**Describe the bug**

Upgrading from OTP 20 to OTP 23 noticed increasing binary memory usage associated with tcp_inet drv_binary allocation type

It is noticeable in the `instrument:allocations()` output:

```
> instrument:allocations()
       ...
       tcp_inet =>
           #{driver_tid => {78,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             driver_tsd => {75,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             drv_binary =>
                 {0,4831,944,64,18,192,140,8,8,15830,0,0,0,0,0,0,0,0},
             drv_internal => {62,80,31,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             port => {0,0,31,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}}}}}

> erlang:memory(binary).
1110936656
```

Another server:

```
> instrument:allocations()
       tcp_inet =>
           #{driver_tid => {79,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             driver_tsd => {76,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             drv_binary =>
                 {7,15020,3852,493,19,114,208,82,32,66378,0,0,0,0,0,0,0,0},
             drv_internal => {62,82,22,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             port => {0,0,22,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}}}}}
        ...
> erlang:memory(binary). 
4555266976
```

A well-behaved node in the same cluster:

```
> instrument:allocations()
       tcp_inet =>
           #{driver_tid => {79,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             driver_tsd => {76,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             drv_binary =>
                 {14,10682,514,65,17,72,212,65,46,2037,0,0,0,0,0,0,0,0},
             drv_internal => {58,87,26,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0},
             port => {0,0,26,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}}}}}

> erlang:memory(binary).
313718152
```

bin_mem_leak_90_days

(The increase on the graph corresponds with upgrading to OTP 23, the flat section previously is OTP 20)

Suspected dist protocol and possibly the 64kb fragmentation logic interacting with refc binaries as when I had tried to dump the core with gdb, the process was suspended long enough to miss the dist tick timeout and it disconnected. When that happened the binary memory was released immediately. 

binary_dealloc_on_dist_disconnection

 (The core was dumped around 21:43 to 21:45)

**To Reproduce**

Out of 6 or so clusters that were upgraded (each with 3 to 12 nodes or so) only observed this on one cluster and on that cluster only on 3 out of 6 nodes.

I couldn't reliable reproduce the issue on my laptop. I had tried simply sending large binaries between two local nodes. Sometimes I see spikes in the  `drv_binary =>` 64kb histogram slot but then it clears out back to 0. 

**Expected behavior**

The binary memory wouldn't keep increasing and would either level off at a low enough level and stay there or get de-allocated.

**Affected versions**

Erlang 23 (erts-11.2.2.10)

```
$ remsh
Erlang/OTP 23 [erts-11.2.2.10] [source] [64-bit] [smp:80:80] [ds:80:80:10] [async-threads:1]
```
```
$ uname -a
Linux db8 4.9.0-17-amd64 #1 SMP Debian 4.9.290-1 (2021-12-12) x86_64 GNU/Linux
```

**Additional context**

vm.args used:
```
-shutdown_time 5000
-kernel error_logger silent
-sasl sasl_error_logger false
+K true
+A32
+zdbbl32768
-ssl session_lifetime 300
+MBacul 0
+MBas aobf
```

I had suspected it was a simple refc binary memory leak and ran `recon:bin_leak(5)` which runs GC for every process but that didn't affect the binary memory.

Other suspicion was that it may be memory fragmentation and had tried running with smaller limit for binary MBC blocks with
```
+MBlmbcs 512
```
And removing `+MBacul 0` and `+MBas aobf`. But that didn't seem to make a difference.

A few more details:

 *  Most of the messages are sent between nodes with `erlang:send(Dest, Msg, [noconnect, nosuspend])` command. In case the messages cannot be sent (response =/= `ok`), there is some buffering with a single head of line sender attempting to re-connect: https://github.com/apache/couchdb/blob/3.x/src/rexi/src/rexi_buffer.erl#L73-L84. Perhaps after Erlang 21 improving `erlang:send/2,3` and making it more asynchronous that buffering strategy is now counter-productive and may be involved in this binary leak issue
 
 * A lot of traffic may be binary data being sent through a coordinator node on the cluster to be written to other nodes. So a common path might be `http socket <-> node1 <-> [dist] <-> node2 <-> disk`, where node1 would be a coordinator and write or read from multiple cluster nodes then respond to the http request
 
 * I had tried to determine the exact size in bytes of the 64kb block by varying the start size of the histogram and the size seems to be `65608`:

```
>f(Hist), {ok, {_, _, #{tcp_inet:=#{drv_binary:=Hist}}}} = instrument:allocations(#{allocator_types=>[binary_alloc], histogram_start=>65609, histogram_width=>4}), Hist.
{78646,2774,0,0}

> f(Hist), {ok, {_, _, #{tcp_inet:=#{drv_binary:=Hist}}}} = instrument:allocations(#{allocator_types=>[binary_alloc], histogram_start=>65608, histogram_width=>4}), Hist.
{58078,23313,0,0}
```

 * Suspecting it may have something to do with dist protocol fragmentation and wanting to mitigate the issue somehow, I had tried to use the release compatibility flag `+R 21` on one of the nodes, but that didn't seem to have any effect.

* I have a gdb core dump and had been trying to inspect it but so far have not gotten very far:

```
(gdb) etp-alloc-instances 10

(gdb) print *((Allctr_t *) 0x7f37742df980)
$2473 = {dd = {q = {tail = {data ...
```

### PR DESCRIPTION
**Pull request by garazdawi**
If a process is suspended while doing a fragmented send and then receives an exit signal it was terminated before it could
finish sending the message leading to a memory leak on the receiving side.

This change fixes so that the message is allowed to finish being sent before the process exits.

Closes #5876

### CODE REVIEW

**Comment by nickva** [comment]
Confirming that this PR fixed the binary leak we had in production: https://github.com/erlang/otp/issues/5876#issuecomment-1100228194

File: erts/emulator/beam/erl_process.c [hunk 1]
```diff
@@ -13897,6 +13900,9 @@ erts_continue_exit_process(Process *p)
         /* notify free */
         erts_atomic32_read_bor_relb(&p->state, ERTS_PSFLG_FREE);
 
+        /* clear suspend count */
+        p->rcount = 0;
```
Comment (garazdawi) [inline erts/emulator/beam/erl_process.c:13904]: Turns out that doing this is not always safe, so need to implement another way to deal with the suspend count.

**Comment by garazdawi** [comment]
The test case needs more work, but I think I'm happy with this solution.

**Comment by garazdawi** [comment]
Fixed the test case to test what it should test.

File: erts/emulator/beam/bif.c [hunk 2]
```diff
@@ -54,7 +54,7 @@ static Export* flush_monitor_messages_trap = NULL;
 static Export* set_cpu_topology_trap = NULL;
 static Export* await_port_send_result_trap = NULL;
 Export* erts_format_cpu_topology_trap = NULL;
-static Export dsend_continue_trap_export;
+Export dsend_continue_trap_export;
```
Comment (nickva) [inline erts/emulator/beam/bif.c:57]: If using a `F_FRAGMENTED_SEND` flag would this still need to be exported or can it remain `static`?

File: erts/emulator/beam/erl_process.c [hunk 3]
```diff
@@ -13945,6 +13955,15 @@ erts_continue_exit_process(Process *p)
         trap_state->pectxt.dist_state = NIL;
         trap_state->pectxt.yield = 0;
 
+        if (p->flags & F_FRAGMENTED_SEND) {
+            /* The process was suspended while doing a fragmented
+               distributed send. We need to finish doing that send
+               as otherwise incomplete fragmented messages will be
+               sent to other nodes potentially causing memory leaks.
+            */
+            trap_state->pectxt.dist_state = p->arg_reg[0];
```
Comment (sverker) [inline erts/emulator/beam/erl_process.c:13964]: Add ASSERT of continuation pointer at trap "dsig_send_continuation".

Comment (sverker) [inline erts/emulator/beam/erl_process.c:13964]: Add ASSERT of continuation pointer at trap "dsig_send_continuation".

**Comment by garazdawi** [comment]
Fix merged to master. It will also be part of the next patches to all currently supported releases.

**Pull request merged**